### PR TITLE
Enhancement/3625 idea hub feature tours on wp pages

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue #3625

## Relevant technical choices

* The only thing that is missing is the storybook. Decided not to include it because it will require recreating the whole WordPress posts page which seems to be overkill.

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/idea-hub/index.js
Comment (aaemnnosttv) [inline assets/js/modules/idea-hub/index.js:63]: Is this because we're loading the Idea Hub module on this page but not Widgets?

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [inline includes/Modules/Idea_Hub.php:189]: `Screen`s use the `googlesitekit-page` class on their render targets, but we shouldn't use that here since we're not rendering a screen. I'm not sure if there's anything actually going in here, but if there is, it should always be wrapped with a `googlesitekit-plugin` class.
```suggestion
					echo '<div id="js-googlesitekit-post-list" class="googlesitekit-plugin"></div>';
```

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [inline includes/Modules/Idea_Hub.php:674]: Is the module needed here? I don't see it used in the tour.

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [inline includes/Modules/Idea_Hub.php:768]: Do we want to add a draft-specific class to all Idea Hub posts? Is this only relevant for drafts or is it for all posts created from IH ideas? Maybe it should be `post` instead or only applied if it is currently a draft?

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [inline includes/Modules/Idea_Hub.php:776]: We should avoid this because the class keeps instance properties for ensuring fonts are only enqueued once. Also, `enqueue_asset` would cause all assets to be registered again for a similar reason.

Without using a `Screen` I think our best option here is to fire an action that the main `Assets` instance can respond to.

Lastly, I think it would be a good idea to separate this into two methods – one for just adding the classes, and another that handles the assets.

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [inline includes/Modules/Idea_Hub.php:762]: Is this function always defined on front end page loads? I feel like it might not be sometimes.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @eugene-manuilov – this is looking good, just a few questions and suggestions for you in addition to some minor things to address.

File: assets/js/components/TourTooltips.js
Comment (eugene-manuilov) [reply assets/js/components/TourTooltips.js:127]: We don't have a ga event category for this feature tour. That's why I needed to add this. Asked Felix to provide it: https://github.com/google/site-kit-wp/issues/3625#issuecomment-895140526

File: assets/js/modules/idea-hub/index.js
Comment (eugene-manuilov) [reply assets/js/modules/idea-hub/index.js:63]: Yes, correct.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php:674]: We check whether the idea hub module is active and connected.

File: assets/js/feature-tours/idea-hub-posts-table.js
Comment (eugene-manuilov) [reply assets/js/feature-tours/idea-hub-posts-table.js]: Updated.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php]: Updated.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php]: Updated to use the `googlesitekit-idea-hub__post` class.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php:767]: Updated to make sure it runs only on in the admin.

File: assets/js/components/Root/index.js
Comment (eugene-manuilov) [reply assets/js/components/Root/index.js:43]: We don't want to display surveys on non-sitekit pages, right? Plus, and we don't need permissions modals when we are not

File: assets/js/googlesitekit-idea-hub-post-list.js
Comment (eugene-manuilov) [reply assets/js/googlesitekit-idea-hub-post-list.js:66]: We pull the current survey and show it in the `CurrentSurveyPortal` component which will not be rendered if `withSurveys` equals `false`. So, answering your question, it will prevent the added survey from showing on this page.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php]: > Without using a Screen I think our best option here is to fire an action that the main Assets instance can respond to.

Added it and updated a few other classes to use the same approach instead of repeating calling `enqueue_fonts()` and `enqueue_asset(...)` methods.

> Lastly, I think it would be a good idea to separate this into two methods – one for just adding the classes, and another that handles the assets.

In this case, these two methods will be 99% identical and will just repeat each other. Not sure whether we really need it. #DRY

File: assets/js/components/Root/index.js
Comment (aaemnnosttv) [reply assets/js/components/Root/index.js:43]: > We don't want to display surveys on non-sitekit pages, right?

Yes, but simply rendering the `<CurrentSurvey>` doesn't trigger a survey which is the more expensive part since it makes a request. Maybe we should make this controllable by a prop but it seems a bit out of scope here without a concrete need.

> Plus, we don't need permissions modals when we are not on a sitekit page.

I'm not sure about that – the permissions modal will be triggered by any insufficient scopes – not only an on-demand request to create something. Again, maybe we should make this conditional but I think we should move this to a new issue to discuss unless needed here.

> These two props help us to turn off unnecessary functionality and avoid a need to load additional assets.

We should be careful to not be too clever to optimize which assets are loaded based on what we know to be used within the component/module. Instead the module should declare it's dependencies and let WP enqueue them as needed. It sounds like maybe you're trying to avoid loading the `core/user` datastore here but I'm not sure we can do that.

File: assets/js/components/TourTooltips.js
Comment (aaemnnosttv) [reply assets/js/components/TourTooltips.js:127]: Thanks – I think we just need to add that to the tour then this condition can be removed. We want to make sure that prop continues to be required.

File: assets/js/googlesitekit-idea-hub-post-list.js
Comment (aaemnnosttv) [reply assets/js/googlesitekit-idea-hub-post-list.js:66]: My mistake, I was referring to the feature tour 🤦‍♂️ 

As mentioned above, I'm not sure these extra props are needed just yet – for surveys in particular they wouldn't be triggered here so I'm not sure rendering the `CurrentSurvey` would be a problem as there wouldn't be one without being triggered anyways 🙂

File: assets/js/components/Root/index.js
Comment (eugene-manuilov) [reply assets/js/components/Root/index.js]: :+1: Reverted.

File: assets/js/components/TourTooltips.js
Comment (eugene-manuilov) [reply assets/js/components/TourTooltips.js]: Added an empty property for now.

File: assets/js/googlesitekit-idea-hub-post-list.js
Comment (eugene-manuilov) [reply assets/js/googlesitekit-idea-hub-post-list.js]: Ok, removed.

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [reply includes/Modules/Idea_Hub.php:768]: Yes of course - what I mean is that `get_current_screen` isn't normally loaded on the front where this filter is also invoked. I see you've added a new guard for the frontend here so this is all good now 👍

File: includes/Core/Util/Activation_Notice.php
Comment (aaemnnosttv) [inline includes/Core/Util/Activation_Notice.php:100]: While this solves our problem, it looks wrong in the context of other usage like this that has access to the `Assets` instance directly.

File: includes/Core/Assets/Assets.php
Comment (aaemnnosttv) [inline includes/Core/Assets/Assets.php:108]: I feel like this shouldn't do more than call `->enqueue_asset( $handle )`, it isn't really clear that this also loads fonts - the problem there of course is that it doesn't currently support loading fonts as a normal asset although I think given the tools we have now we could make that possible (the problem here is that it uses a style for AMP and a script for web). I don't want to increase the scope of this issue any farther though. I suppose we could go with this for now and then come back to it in a separate issue.

One alternative would be to provide the `Assets` instance to module classes but I don't see a great way to do that.

Thoughts @felixarntz ?

File: includes/Core/Assets/Assets.php
Comment (aaemnnosttv) [inline includes/Core/Assets/Assets.php:103]: This may interfere with the action-related checks in `enqueue_fonts`.

File: assets/js/modules/idea-hub/index.js
Comment (aaemnnosttv) [reply assets/js/modules/idea-hub/index.js:63]: See my comment regarding `googlesitekit-modules-idea-hub` – we shouldn't need to load it so this would also be unnecessary I think in that case.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @eugene-manuilov – I left a few more comments/replies for you. I'd also like to get @felixarntz 's thoughts on the change to make Assets available to the Idea Hub module. See my comments below.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php]: Yes, good catch. Updated.

File: assets/js/components/Root/index.js
Comment (aaemnnosttv) [inline assets/js/components/Root/index.js:65]: This looks like it can be reverted as there there should never be a case where we render `Root` without children.

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [inline includes/Modules/Idea_Hub.php:192]: Let's scope this to the specific screen rather than all admin screens.
```suggestion
				'admin_footer-edit.php',
```

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [inline includes/Modules/Idea_Hub.php:194]: Let's also only output this for posts.
```suggestion
					if ( 'post' !== get_current_screen()->post_type ) {
						return;
					}
					echo '<div id="js-googlesitekit-post-list" class="googlesitekit-plugin"></div>';
```

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [inline includes/Modules/Idea_Hub.php:773]: Let's use the screen object for this instead.
```suggestion
```

File: includes/Modules/Idea_Hub.php
Comment (aaemnnosttv) [inline includes/Modules/Idea_Hub.php:780]:
```suggestion
		if ( is_null( $current_screen ) || 'edit-post' !== $current_screen->id || 'post' !== $current_screen->post_type ) {
```

**Review by aaemnnosttv — COMMENTED** [review]
@eugene-manuilov A few final tweaks to make here apart from the changes around Assets pending review by @felixarntz .

**Comment by eugene-manuilov** [comment]
@googlebot I fixed it.

**Comment by eugene-manuilov** [comment]
@googlebot I fixed it.

**Comment by eugene-manuilov** [comment]
:rage1: @googlebot I fixed it.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php]: Updated.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php]: Updated.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php]: Updated.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [reply includes/Modules/Idea_Hub.php]: Updated.

File: assets/js/components/Root/index.js
Comment (eugene-manuilov) [reply assets/js/components/Root/index.js:65]: We have the `Root` element without children here: https://github.com/google/site-kit-wp/blob/9fe10cb8c8214391288ba078e5fa4520cca52f36/assets/js/googlesitekit-idea-hub-post-list.js#L61-L66

File: includes/Core/Assets/Assets.php
Comment (eugene-manuilov) [reply includes/Core/Assets/Assets.php]: Updated to use $assets instance created in the Plugins class.

File: includes/Plugin.php
Comment (aaemnnosttv) [inline includes/Plugin.php:171]: I see why it's necessary now but we should avoid this. It doesn't need to block this PR though and I can open a follow-up to fix it.

**Review by aaemnnosttv — APPROVED** [review]
Great, thanks @eugene-manuilov !

**Pull request merged**
